### PR TITLE
Switch to C++ 20 (partially)

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Small changes needed to be C++20 compatible
- Switch to C++20 as the standard
- We can't use all C++20 features because compilers haven't implemented them all yet. However, with our current requirements most features will be available.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: cmake/SetCxxStandard.cmake [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Distributed under the MIT License.
 # See LICENSE.txt for details.
 
-set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD 20)
```
Comment (knelli2) [inline cmake/SetCxxStandard.cmake:4]: Have you tried this on all the clusters?

Comment (nilsdeppe) [reply cmake/SetCxxStandard.cmake:4]: I have not, though this depends only on the available/required compilers which we do test in CI :shrug:

Comment (nilsvu) [inline cmake/SetCxxStandard.cmake:4]: Want to update the standard badge in the readme as well?

**Review by nilsvu — DISMISSED** [review]

**Comment by wthrowe** [comment]
I have no additional comments.

I have tested it on my machine with clang 17 and all the tests pass.

**Comment by nilsvu** [comment]
Same on AppleClang 15, just a couple of warnings:

```
spectre/tests/Unit/Domain/Structure/Test_Direction.cpp:107:3: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
  std::is_sorted(Direction<1>::all_directions().begin(),
  ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
spectre/tests/Unit/Domain/Structure/Test_Direction.cpp:109:3: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
  std::is_sorted(Direction<2>::all_directions().begin(),
  ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
spectre/tests/Unit/Domain/Structure/Test_Direction.cpp:111:3: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
  std::is_sorted(Direction<3>::all_directions().begin(),
  ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 warnings generated.

spectre/tests/Unit/Utilities/ErrorHandling/Test_FloatingPointExceptions.cpp:32:45: warning: unused function 'throw_invalid' [-Wunused-function]
[[noreturn]] __attribute__((noinline)) void throw_invalid() {
                                            ^
spectre/tests/Unit/Utilities/ErrorHandling/Test_FloatingPointExceptions.cpp:40:45: warning: unused function 'throw_overflow' [-Wunused-function]
[[noreturn]] __attribute__((noinline)) void throw_overflow() {
                                            ^
spectre/tests/Unit/Utilities/ErrorHandling/Test_FloatingPointExceptions.cpp:48:45: warning: unused function 'throw_div_by_zero' [-Wunused-function]
[[noreturn]] __attribute__((noinline)) void throw_div_by_zero() {
```

**Comment by knelli2** [comment]
I get these linking errors when I build the unit tests on Wheeler: [tests.txt](https://github.com/sxs-collaboration/spectre/files/14226459/tests.txt)

And these linking errors when I build the `AnalyticTestCharacteristicExtract` executable on Wheeler: [rot_schw.txt](https://github.com/sxs-collaboration/spectre/files/14226469/rot_schw.txt)

**Comment by wthrowe** [comment]
> Same on AppleClang 15, just a couple of warnings:
> 
> ```
> spectre/tests/Unit/Domain/Structure/Test_Direction.cpp:107:3: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
>   std::is_sorted(Direction<1>::all_directions().begin(),
>   ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> spectre/tests/Unit/Domain/Structure/Test_Direction.cpp:109:3: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
>   std::is_sorted(Direction<2>::all_directions().begin(),
>   ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> spectre/tests/Unit/Domain/Structure/Test_Direction.cpp:111:3: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
>   std::is_sorted(Direction<3>::all_directions().begin(),
>   ^~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 3 warnings generated.
> 

These are actual existing problems.  (Not related to C++20.)

> spectre/tests/Unit/Utilities/ErrorHandling/Test_FloatingPointExceptions.cpp:32:45: warning: unused function 'throw_invalid' [-Wunused-function]
> [[noreturn]] __attribute__((noinline)) void throw_invalid() {
>                                             ^
> spectre/tests/Unit/Utilities/ErrorHandling/Test_FloatingPointExceptions.cpp:40:45: warning: unused function 'throw_overflow' [-Wunused-function]
> [[noreturn]] __attribute__((noinline)) void throw_overflow() {
>                                             ^
> spectre/tests/Unit/Utilities/ErrorHandling/Test_FloatingPointExceptions.cpp:48:45: warning: unused function 'throw_div_by_zero' [-Wunused-function]
> [[noreturn]] __attribute__((noinline)) void throw_div_by_zero() {
> ```

These are some stuff that should be `#ifdef`ed out for apple.

**Comment by nilsdeppe** [comment]
I've pushed commits that fix the things noted in the comments. I tested clang debug/release and GCC debug on Wheeler and those build and tests pass.

File: cmake/SetCxxStandard.cmake (hunk 1 above)
Comment (nilsvu) [reply cmake/SetCxxStandard.cmake:4]: Ping

**Review by nilsvu — DISMISSED** [review]

**Review by knelli2 — DISMISSED** [review]
Haven't tested on any other clusters, but I think we can fix problems as they arise. As long as Wheeler is working, that should be good for now to not interrupt people's workflow.

**Comment by nilsdeppe** [comment]
Rebased and updated the README badge to 20

**Review by nilsvu — APPROVED** [review]

**Pull request merged**